File: linked-list-easy/203.移除链表元素.py
#
# @lc app=leetcode.cn id=203 lang=python3
#
# [203] 移除链表元素
#

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# @lc code=start

# 执行用时： 72 ms , 在所有 Python3 提交中击败了 60.10% 的用户
# 内存消耗： 17.8 MB , 在所有 Python3 提交中击败了 69.90% 的用户
class Solution:
    def removeElements(self, head: ListNode, val: int) -> ListNode:
        # For situation like: [2], 2
        dummy = ListNode(0, head)
        pre, cur = dummy, head
        while cur is not None:
            if cur.val == val:
                pre.next = cur.next
            else:
                pre = cur
            cur = cur.next
        # For situation like: [2], 2
        return dummy.next


# @lc code=end

File: linked-list-easy/203.移除链表元素.java
/*
 * @lc app=leetcode.cn id=203 lang=java
 *
 * [203] 移除链表元素
 */

/**
 * Definition for singly-linked list.
 */

public class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

// @lc code=start

// 65/65 cases passed (1 ms)
// Your runtime beats 99.68 % of java submissions
// Your memory usage beats 10.67 % of java submissions (39.7 MB)
class Solution {
    /* 递归 */
    public ListNode removeElements(ListNode head, int val) {
        if (head == null) {
            return head;
        }
        head.next = removeElements(head.next, val);
        if (head.val == val) {
            return head.next;
        } else {
            return head;
        }
    }
}

// @lc code=end



class Solution_1 {
    /* 删除头结点时另做考虑 */
    public ListNode removeElements(ListNode head, int val) {
        //删除值相同的头结点后，可能新的头结点也值相等，用循环解决
        while (head != null && head.val == val) {
            head = head.next;
        }
        if (head == null)
            return head;
        ListNode prev = head;
        //确保当前结点后还有结点
        while(prev.next != null) {
            if (prev.next.val == val) {
                prev.next = prev.next.next;
            } else {
                prev = prev.next;
            }
        }
        return head;
    }
}
